pkg/controller/storageprofile-controller.go

```diff
@@ -0,0 +1,201 @@
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			storageProfile, err = r.createEmptyStorageProfile(sc)
+		} else {
+			log.Error(err, "Unable to create StorageProfile")
+			return reconcile.Result{}, err
+		}
+	} else {
+		prevStorageProfile = storageProfile.DeepCopyObject()
+	}
+
+	storageProfile.Status.StorageClass = &sc.Name
+	storageProfile.Status.Provisioner = &sc.Provisioner
+	storageProfile.Status.ClaimPropertySets = storageProfile.Spec.ClaimPropertySets
+
+	var claimPropertySet *cdiv1.ClaimPropertySet
+	// right now the CDI supports only a single property set
+	if len(storageProfile.Status.ClaimPropertySets) > 0 {
+		claimPropertySet = &storageProfile.Status.ClaimPropertySets[0]
+	} else {
+		claimPropertySet = &cdiv1.ClaimPropertySet{}
+	}
+
+	r.reconcileAccessModes(sc, claimPropertySet)
+	r.reconcileVolumeMode(sc, claimPropertySet)
+
+	if !isClaimPropertySetEmpty(claimPropertySet) {
+		storageProfile.Status.ClaimPropertySets = []cdiv1.ClaimPropertySet{*claimPropertySet}
+	}
+
+	if prevStorageProfile == nil {
+		if err := r.client.Create(context.TODO(), storageProfile); err != nil {
+			return reconcile.Result{}, err
+		}
+	} else if !reflect.DeepEqual(prevStorageProfile, storageProfile) {
+		// Updates have happened, update StorageProfile.
+		log.Info("Updating StorageProfile", "StorageProfile.Name", storageProfile.Name, "storageProfile", storageProfile)
+		if err := r.client.Update(context.TODO(), storageProfile); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func isClaimPropertySetEmpty(set *cdiv1.ClaimPropertySet) bool {
+	return set == nil ||
+		(len(set.AccessModes) == 0 && set.VolumeMode == nil)
+}
+
+func (r *StorageProfileReconciler) reconcileVolumeMode(sc *storagev1.StorageClass, claimPropertySet *cdiv1.ClaimPropertySet) {
+	if claimPropertySet.VolumeMode == nil {
+		capabilities, found := storagecapabilities.Get(sc)
+		if found {
+			claimPropertySet.VolumeMode = &capabilities.VolumeMode
+		}
+	}
+}
+
+func (r *StorageProfileReconciler) reconcileAccessModes(sc *storagev1.StorageClass, claimPropertySet *cdiv1.ClaimPropertySet) {
+	// reconcile accessModes
+	if len(claimPropertySet.AccessModes) == 0 {
+		capabilities, found := storagecapabilities.Get(sc)
+		if found {
+			claimPropertySet.AccessModes = []v1.PersistentVolumeAccessMode{capabilities.AccessMode}
+		}
+	}
+}
+
+func (r *StorageProfileReconciler) getStorageProfile(sc *storagev1.StorageClass) (*cdiv1.StorageProfile, error) {
+	storageProfile := &cdiv1.StorageProfile{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: sc.Name}, storageProfile); err != nil {
+		return nil, err
+	}
+	return storageProfile, nil
+}
+
+func (r *StorageProfileReconciler) createEmptyStorageProfile(sc *storagev1.StorageClass) (*cdiv1.StorageProfile, error) {
+	storageProfile := MakeEmptyStorageProfileSpec(sc.Name)
+	if err := operator.SetOwnerRuntime(r.uncachedClient, storageProfile); err != nil {
```

Comment:
I suggest you put a comment in here why you are using the uncached client, so I stop asking why you are using it.

Reply:
you are totally right, because now I need to find-out again why do I use it ;)

Reply:
DONE